src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -78,124 +67,251 @@ public static IEnumerable<VisualStudioInstance> GetInstances(string workingDirec
             }
         }
 
-        /// <summary>
-        /// This native method call determines the actual location of path, including
-        /// resolving symbolic links.
-        /// </summary>
-        private static string realpath(string path)
+        private static IEnumerable<string> GetDotNetBasePaths(string workingDirectory)
         {
-            IntPtr ptr = NativeMethods.realpath(path, IntPtr.Zero);
-            string result = Marshal.PtrToStringAuto(ptr);
-            NativeMethods.free(ptr);
-            return result;
+            LoadHostFxr();
+
+            string? bestSDK = GetSdkFromGlobalSettings(workingDirectory);
+            if (!string.IsNullOrEmpty(bestSDK))
+                yield return bestSDK;
+
+            string[] dotnetPaths = GetAllAvailableSDKs();
+            // We want to return the newest SDKs first, however, so iterate over the list in reverse order.
+            // If basePath is disqualified because it was later
+            // than the runtime version, this ensures that RegisterDefaults will return the latest valid
+            // SDK instead of the earliest installed.
+            for (int i = dotnetPaths.Length - 1; i >= 0; i--)
+            {
+                if (dotnetPaths[i] != bestSDK)
+                    yield return dotnetPaths[i];
+            }
+        }
+        private static void LoadHostFxr()
+        {
+            var isOSX = RuntimeInformation.IsOSPlatform(OSPlatform.OSX);
+            if (isOSX)
+            {
+                var loadContext = System.Runtime.Loader.AssemblyLoadContext.GetLoadContext(Assembly.GetExecutingAssembly());
+                if (loadContext != null)
+                    loadContext.ResolvingUnmanagedDll += HostFxrResolver;
+            }
         }
 
-        private static string FindDotnetFromEnvironmentVariable(string environmentVariable, string exeName)
+        private static IntPtr HostFxrResolver(Assembly assembly, string libraryName)
```

Comment:
Why do you have separate LoadHostFxr(string) and LoadHostFxr() methods? From what I see, the former is never used.

Reply:
removed. I will configure code styles in a separate PR to avoid these dangling things.